Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,324 @@
+    /// the `Error`. In the case of OAuth2, if you authentication server can invalidate access tokens, then you need
+    /// to use this method to inspect the `HTTPURLResponse` looking for an indicator that this happened. This is
+    /// commonly handled by the authentication server returning a 401 and some additional header to indicate that
+    /// the `Authorization` header is invalid or not authorized.
+    ///
+    /// It is very important to understand how your authentication server works to be able to implement this correctly.
+    /// For example, if your authentication server returns a 401 when an OAuth2 error occurs, and your downstream
+    /// service also returns a 401 when you are not authorized to perform that operation, how do you know which layer
+    /// of the backend returned you a 401? You do not want to trigger a refresh unless you know your authentication
+    /// server is actually the layer rejecting the request.
+    ///
+    /// If your authentication server will never reject non-expired credentials for performance reasons, then you can
+    /// safely return `false` for this method.
+    func didRequest(_ urlRequest: URLRequest, with response: HTTPURLResponse, failDueToAuthenticationError error: Error) -> Bool
+
+    /// NOTE: CN rework docstring
+    ///
+    /// Returns whether the `URLRequest` was authenticated with the specified `Credential`. If it was not, then we know
+    /// the `URLRequest` was authenticated with a previous `Credential` and can be immediately retried with the new
+    /// `Credential`.
+    ///
+    /// This is an edge case that can occur if the authentication server can invalidate non-expired credentials. In
+    /// this case, it's possible for requests to be executed with non-expired, invalid credentials. Refresh will be
+    /// triggered when the first failing response comes back from the authentication server. It is possible that some
+    /// of the requests will not return from the authentication server until the refresh is completed. In this case,
+    /// we inspect the slower requests when returned from the authentication server to see if the credential applied
+    /// to them is the same as the latest credential. If not, then we know we've refreshed the credential while the
+    /// request was in flight and it is immediately retried with the new credential. If yes, then we know we've found
+    /// the first request to fail due to a non-expired, invalidate credential and refresh is triggered.
+    func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: AuthCredential) -> Bool
+}
+
+// MARK: -
+
+/// NOTE: CN docstring
+public enum AuthenticationError: Error {
+    /// NOTE: CN docstring
+    case missingCredential
+
+    /// NOTE: CN docstring
+    case excessiveRefresh
+}
+
+// MARK: -
+
+/// NOTE: CN docstring
+public class AuthenticationInterceptor<AuthenticatorType, AuthCredential>: RequestInterceptor where
```

Comment:
Do we need the `AuthCredential` type? Wouldn't `AuthenticatorType.AuthCredential` work just as well?

Reply:
Yep, good call. Addressed in d7b4d1d1.